lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -256,6 +327,9 @@ unsigned RecursiveTypeProperties::toFlags() const {
 // FIRRTLType Implementation
 //===----------------------------------------------------------------------===//
 
+/// Print this type.
+void FIRRTLType::print(DialectAsmPrinter &os) const { printType(*this, os); }
```

Comment:
The whole dispatch path here seems a bit complicated, can we just delete this function?

Reply:
Or we could move the body of `customTypePrinter(Type type, DialectAsmPrinter &os)` here, since it handles only the FIRRTLType types.

Reply:
That's how it originally was. But you can use TableGen to come up with new descendants of `FIRRTLType`, in which case you'd have to call the generated printer even for a `FIRRTLType`. That might be attractive for some of the nullary types like `clock` or `reset`. I tried to disentangle the fact that a type is generated from whether it inherits from `FIRRTLType` or not, because you can have both:
```
FIRRTLDialect::printType   FIRRTLType::print
           │                     │
           └─────────┐   ┌───────┘
                     ▼   ▼
                static printType
                       │
            ┌──────────┴──────────┐
            ▼                     ▼
   generatedTypePrinter   customTypePrinter
```

Reply:
Thanks for the explanation.  I guess back to my first idea: `FIRRTLType::print(DialectAsmPrinter &os)` is not really useful to have in the public interface of `FIRRTLType`, and we could switch all existing calls to `::printType` e.g.
```c++
vectorType.getElementType().print(os);
// to 
::printType(vectorType.getElementType());
```

Reply:
Yeah that would definitely be an option. On the other hand, the tablegen'd types all have a ```void print(::mlir::DialectAsmPrinter &printer) const;``` in there, so maybe there's an expectation for a type to provide a `print()` function? For example, if you have a `FIRRTLType` anywhere else in the code base, outside `FIRRTLTypes.cpp`, wouldn't you want to be able to print that with something like `x.print(...)` somehow? Or is there a different way to do that?

Reply:
I don't know how to get a `DialectAsmPrinter`, I was thinking this is an internal thing to the printing framework?  The normal way I think is to go through[ `Type::print(raw_ostream &os)`](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Types.h#L158)

Reply:
Ah good point, you're totally right! And I suspect that `print(raw_ostream&)` then dispatches through the dialect interface anyway.

Reply:
Yeah that goes through the [`AsmPrinter`](https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/AsmPrinter.cpp#L2735), so I guess we're coming from the dialect side anyway.

Reply:
I guess if you wanted to mix FIRRTL types into other dialect types and skip printing the `!firrtl.` on the front of the type (e.g. `!sv.vector<bundle<a: uint>>`) , this could be useful, is that the kind of use case you're thinking of?  Should we have a matching `parse` function for that to work?

Reply:
Woops github didn't update before i sent the previous comment. You can ignore it!

Reply:
:smile: I removed `FIRRTLType::print` and replaced the uses (of which there were only the ones in `FIRRTLTypes.cpp` as you pointed out) with the static `printType`. Works like a charm. Also, all the dialect prefixing `!firrtl.*` still seems to work -- looks like that's handled as part of the `DialectAsmPrinter`. There must be some dialect context kept in the printer, because it seems to drop that prefix based on context.